File: app/src/main/java/com/kim344/retrofitretry/java/RetrofitInit.java
package com.kim344.retrofitretry.java;

import android.util.Log;

import java.security.SecureRandom;
import java.security.cert.CertificateException;
import java.security.cert.X509Certificate;
import java.util.concurrent.TimeUnit;

import javax.net.ssl.HostnameVerifier;
import javax.net.ssl.SSLContext;
import javax.net.ssl.SSLSession;
import javax.net.ssl.SSLSocketFactory;
import javax.net.ssl.TrustManager;
import javax.net.ssl.X509TrustManager;

import okhttp3.OkHttpClient;
import okhttp3.logging.HttpLoggingInterceptor;
import retrofit2.Retrofit;
import retrofit2.adapter.rxjava2.RxJava2CallAdapterFactory;
import retrofit2.converter.gson.GsonConverterFactory;

import static com.kim344.retrofitretry.java.Constant.TIMEOUT_SECOND;
import static com.kim344.retrofitretry.java.Constant.URL_HOST;

public class RetrofitInit {

    private static RetrofitInit instance;

    RetrofitInit getInstance(){
        if (instance != null) {}
        else instance = new RetrofitInit();
        return instance;
    }

    private Retrofit retrofit = new Retrofit.Builder()
            .baseUrl(URL_HOST + "/")
            .client(unSafeOkHttpClient())
            .addCallAdapterFactory(RxJava2CallAdapterFactory.create())
            .addConverterFactory(GsonConverterFactory.create())
            .build();

    private Service service = retrofit.create(Service.class);

    Service getService(){
        if (service != null) {}
        else service = retrofit.create(Service.class);
        return service;
    }

    private OkHttpClient unSafeOkHttpClient(){
        try {
            final TrustManager[] trustAllCerts = new TrustManager[]{
                    new X509TrustManager() {
                        @Override
                        public void checkClientTrusted(X509Certificate[] x509Certificates, String s) throws CertificateException {

                        }

                        @Override
                        public void checkServerTrusted(X509Certificate[] x509Certificates, String s) throws CertificateException {

                        }

                        @Override
                        public X509Certificate[] getAcceptedIssuers() {
                            return new X509Certificate[0];
                        }
                    }
            };

            SSLContext sslContext = SSLContext.getInstance("SSL");
            sslContext.init(null,trustAllCerts,new SecureRandom());

            SSLSocketFactory sslSocketFactory = sslContext.getSocketFactory();

            return new OkHttpClient.Builder()
                    .connectTimeout(TIMEOUT_SECOND, TimeUnit.SECONDS)
                    .writeTimeout(TIMEOUT_SECOND,TimeUnit.SECONDS)
                    .readTimeout(TIMEOUT_SECOND,TimeUnit.SECONDS)
                    .addInterceptor(loggingInterceptor())
                    .sslSocketFactory(sslSocketFactory,(X509TrustManager) trustAllCerts[0])
                    .hostnameVerifier(new HostnameVerifier() {
                        @Override
                        public boolean verify(String s, SSLSession sslSession) {
                            return true;
                        }
                    }).build();

        } catch (Exception e){
            throw new RuntimeException(e);
        }
    }

    private HttpLoggingInterceptor loggingInterceptor() {
        HttpLoggingInterceptor interceptor = new HttpLoggingInterceptor(new HttpLoggingInterceptor.Logger() {
            @Override
            public void log(String message) {
                Log.e("Retrofit :", message + "");
            }
        });
        interceptor.setLevel(HttpLoggingInterceptor.Level.BODY);
        return interceptor;
    }

}

File: app/src/main/java/com/kim344/retrofitretry/kotlin/RetrofitInit.kt
package com.kim344.retrofitretry.kotlin

import android.util.Log
import com.kim344.retrofitretry.java.RetrofitInit
import com.kim344.retrofitretry.java.Service
import com.kim344.retrofitretry.kotlin.Constant.TIMEOUT_SECOND
import com.kim344.retrofitretry.kotlin.Constant.URL_HOST
import okhttp3.OkHttpClient
import okhttp3.logging.HttpLoggingInterceptor
import retrofit2.Retrofit
import retrofit2.adapter.rxjava2.RxJava2CallAdapterFactory
import retrofit2.converter.gson.GsonConverterFactory
import java.security.SecureRandom
import java.security.cert.CertificateException
import java.security.cert.X509Certificate
import java.util.concurrent.TimeUnit
import javax.net.ssl.*

class RetrofitInit {
    private var instance: RetrofitInit? = null

    internal fun getInstance(): RetrofitInit {
        if (instance != null) {
        } else
            instance = RetrofitInit()
        return instance as RetrofitInit
    }

    private val retrofit = Retrofit.Builder()
        .baseUrl("$URL_HOST/")
        .client(unSafeOkHttpClient())
        .addCallAdapterFactory(RxJava2CallAdapterFactory.create())
        .addConverterFactory(GsonConverterFactory.create())
        .build()

    private var service: Service? = retrofit.create(Service::class.java)

    internal fun getService(): Service? {
        if (service != null) {
        } else
            service = retrofit.create(Service::class.java)
        return service
    }

    private fun unSafeOkHttpClient(): OkHttpClient {
        try {
            val trustAllCerts = arrayOf<TrustManager>(object : X509TrustManager {
                @Throws(CertificateException::class)
                override fun checkClientTrusted(
                    x509Certificates: Array<X509Certificate>,
                    s: String
                ) {

                }

                @Throws(CertificateException::class)
                override fun checkServerTrusted(
                    x509Certificates: Array<X509Certificate>,
                    s: String
                ) {

                }

                override fun getAcceptedIssuers(): Array<X509Certificate?> {
                    return arrayOfNulls(0)
                }
            })

            val sslContext = SSLContext.getInstance("SSL")
            sslContext.init(null, trustAllCerts, SecureRandom())

            val sslSocketFactory = sslContext.socketFactory

            return OkHttpClient.Builder()
                .connectTimeout(TIMEOUT_SECOND.toLong(), TimeUnit.SECONDS)
                .writeTimeout(TIMEOUT_SECOND.toLong(), TimeUnit.SECONDS)
                .readTimeout(TIMEOUT_SECOND.toLong(), TimeUnit.SECONDS)
                .addInterceptor(loggingInterceptor())
                .sslSocketFactory(sslSocketFactory, trustAllCerts[0] as X509TrustManager)
                .hostnameVerifier { s, sslSession -> true }.build()

        } catch (e: Exception) {
            throw RuntimeException(e)
        }

    }

    private fun loggingInterceptor(): HttpLoggingInterceptor {
        val interceptor = HttpLoggingInterceptor(HttpLoggingInterceptor.Logger { message ->
            Log.e(
                "Retrofit :",
                message + ""
            )
        })
        interceptor.level = HttpLoggingInterceptor.Level.BODY
        return interceptor
    }
}